packages/protocol/contracts/layer1/preconf/impl/Overseer.sol

```diff
@@ -0,0 +1,289 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.24;
+
+import "src/layer1/preconf/iface/IOverseer.sol";
+import "src/layer1/preconf/libs/LibPreconfConstants.sol";
+import "openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+import "src/shared/common/EssentialContract.sol";
+import "@solady/src/utils/MerkleProofLib.sol";
+import "@solady/src/utils/MerkleTreeLib.sol";
+
+/// @title Overseer
+/// @notice The Overseer is responsible for blacklisting validators of preconf operators based on
+/// subjective faults.
+/// For instance, non-adherence to fair exchange.
+/// @dev Operators of blacklisted validators are not inserted in the lookahead. This is done to
+/// prevent the lookahead from being polluted by invalidators.
+/// @custom:security-contact security@taiko.xyz
+contract Overseer is IOverseer, EssentialContract {
+    uint64 public signingThreshold;
+    uint64 public numSigners;
+    uint64 public nonce;
+
+    mapping(address signerAddress => bool isSigner) public signers;
+
+    /// @dev Maps the root of a merkle tree of validator keys to the timestamp at which they
+    /// were blacklisted or unblacklisted.
+    mapping(bytes32 validatorsRoot => BlacklistTimestamps blacklistTimestamps) public blacklist;
+
+    uint256[47] private __gap;
+
+    function init(uint64 _signingThreshold, address[] memory _signers) external initializer {
+        __Essential_init(address(0));
+        __OverseerInit(_signingThreshold, _signers);
+    }
+
+    // Blacklist functions
+    // -----------------------------------------------------------------------------------
+
+    /// @inheritdoc IOverseer
+    function blacklistValidators(
+        BLS.G1Point[] calldata _validatorPubKeys,
+        bytes[] memory _signatures
+    )
+        external
+    {
+        // Merkleize the validators to a single validator's merkle root
```

Comment:
How much overhead do we have with Merkalizing the validator pub keys? I expect blacklisting to be done on a per-validator basis rather in bulk for most cases, so if the overhead is large we can consider blacklisting per validator.

Reply:
Hmm, rethinking, I think it makes sense to blacklist the whole validator set that is assigned to the faulty operator key, which I think you were indicating from the start.